src/array/cpu/rowwise_pick.h

```diff
@@ -132,6 +158,132 @@ COOMatrix CSRRowWisePick(CSRMatrix mat, IdArray rows,
                    picked_row, picked_col, picked_idx);
 }
 
+// Template for picking non-zero values row-wise. The implementation utilizes
+// OpenMP parallelization on rows because each row performs computation independently.
+template <typename IdxType>
+COOMatrix CSRRowWisePerEtypePick(CSRMatrix mat, IdArray rows, IdArray etypes,
+                                 int64_t num_picks, bool replace, RangePickFn<IdxType> pick_fn) {
+  using namespace aten;
+  const IdxType* indptr = static_cast<IdxType*>(mat.indptr->data);
+  const IdxType* indices = static_cast<IdxType*>(mat.indices->data);
+  const IdxType* data = CSRHasData(mat)? static_cast<IdxType*>(mat.data->data) : nullptr;
+  const IdxType* rows_data = static_cast<IdxType*>(rows->data);
+  const int32_t* etype_data = static_cast<int32_t*>(etypes->data);
+  const int64_t num_rows = rows->shape[0];
+  const auto& ctx = mat.indptr->ctx;
+  CHECK_EQ(etypes->dtype.bits / 8, sizeof(int32_t));
+  std::vector<IdArray> picked_rows(rows->shape[0]);
+  std::vector<IdArray> picked_cols(rows->shape[0]);
+  std::vector<IdArray> picked_idxs(rows->shape[0]);
+
+  for (int64_t i = 0; i < num_rows; ++i) {
+    const IdxType rid = rows_data[i];
+    CHECK_LT(rid, mat.num_rows);
+    const IdxType off = indptr[rid];
+    const IdxType len = indptr[rid + 1] - off;
+
+    // do something here
+    if (len == 0) {
+      picked_rows[i] = NewIdArray(0, ctx, sizeof(IdxType) * 8);
+      picked_cols[i] = NewIdArray(0, ctx, sizeof(IdxType) * 8);
+      picked_idxs[i] = NewIdArray(0, ctx, sizeof(IdxType) * 8);
+      continue;
+    }
+
+    // fast path
+    if (len <= num_picks && !replace) {
+      IdArray rows = Full(rid, len, sizeof(IdxType) * 8, ctx);
+      IdArray cols = Full(-1, len, sizeof(IdxType) * 8, ctx);
+      IdArray idx = Full(-1, len, sizeof(IdxType) * 8, ctx);
+      IdxType* cdata = static_cast<IdxType*>(cols->data);
+      IdxType* idata = static_cast<IdxType*>(idx->data);
+      for (int64_t j = 0; j < len; ++j) {
+        cdata[j] = indices[off + j];
+        idata[j] = data ? data[off + j] : off + j;
```

Comment:
if data is null, what is filled in `idata`?

Reply:
The eid (off + j)